File: yaid-ts/yaid.ts
import { base32Encode, base32Decode } from "./crockford.js";

const TIME_BYTES = 5;
const DIFF_BYTES = 2;
const META_BYTES = 1;

const SIZE = TIME_BYTES + DIFF_BYTES + META_BYTES;

// 1 for miliseconds, 10 for centiseconds, 1000 for seconds, etc.
const DEFIDER = 10;

export const MAX_TIMESTAMP = 1099511627775;

const c: Crypto = globalThis.crypto || crypto;

/*
// NodeJS < 20 does not support globalThis.
// Import node.js webcrypto
if (!c) {
    c = require("crypto").webcrypto;
}
*/

export class YAID {
	constructor(private bytes = new Uint8Array(SIZE)) {
		if (bytes.byteLength != SIZE) {
			throw new RangeError("bytes length must be " + SIZE + " not " + bytes.byteLength);
		}
	}

	differentiator(): Uint8Array {
		return this.bytes.slice(TIME_BYTES, TIME_BYTES + DIFF_BYTES);
	}

	setDifferentiator(d: Uint8Array) {
		this.bytes.set(d, TIME_BYTES);
	}

	meta(): number {
		return this.bytes[TIME_BYTES + DIFF_BYTES];
	}

	/**
	 * Set meta data
	 * @param number between 0 and 255
	 */
	setMeta(i: number) {
		if (i < 0 || i > 255) {
			throw new RangeError("meta must be within 0-255");
		}
		this.bytes[TIME_BYTES + DIFF_BYTES] = i;
	}

	time(): Date {
		const ms = this.timestamp() * DEFIDER;
		return new Date(ms);
	}

	setTime(t: Date) {
		const cs = t.getTime() / DEFIDER;
		this.setTimestamp(cs);
	}

	/** Time as hundredth of a second since January 1, 1970 12:00:00 AM UTC */
	timestamp(): number {
		return (
			this.bytes[4] |
			(this.bytes[3] << 8) |
			(this.bytes[2] << 16) |
			(this.bytes[1] << 24) |
			(this.bytes[0] << 32)
		);
	}

	/** Set the time as hundredth of a second since January 1, 1970 12:00:00 AM UTC */
	setTimestamp(t: number) {
		if (t > MAX_TIMESTAMP) {
			throw new Error("timestamp must not be greater than " + MAX_TIMESTAMP);
		}
		/*
		this.bytes[0] = t >> 32;
		this.bytes[1] = (t >> 24) & 255;
		this.bytes[2] = (t >> 16) & 255;
		this.bytes[3] = (t >> 8) & 255;
		this.bytes[4] = t & 255;
		*/
		for (let i = 0, j = 1; i < TIME_BYTES; i++, j *= 0x100) {
			this.bytes[TIME_BYTES - 1 - i] = (t / j) & 0xff;
		}
	}

	toBytes(): Uint8Array {
		return this.bytes;
	}

	toString(): string {
		return base32Encode(this.bytes);
	}
}

/**
 * New YAID
 * @param meta
 * @param size Size of the id in bytes
 * @param time
 * @returns YAID
 */
export function New(meta?: number, size = 8, time = new Date()): YAID {
	if (typeof meta === "undefined") {
		meta = random(META_BYTES)[0];
	}

	const y = new YAID();
	y.setTime(time);
	y.setDifferentiator(random(DIFF_BYTES));
	y.setMeta(meta);

	return y;
}

export function Parse(yaid: string): YAID {
	const b = base32Decode(yaid);
	return new YAID(b);
}

// Return n bytes of random data
function random(n: number): Uint8Array {
	const m = new Uint8Array(n);
	return c.getRandomValues(m);
}

File: yaid-go/yaid.go
// Copyright (c) 2023 Hans van Leeuwen. MIT Licensed. See LICENSE.md for full license.

/*
Package yaid is a go package that implements [Yet Another ID]

# Example

	import (
		"github.com/hnz/yaid/yaid-go"
	)

	y := yaid.NewGenerator()
	id := y()

[Yet Another ID]: https://github.io/yaid
*/
package yaid

import (
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"io"
	"time"

	"github.com/ilius/crock32"
)

// Get the maximum timestamp by setting all bytes to maximum value
var maxTimestamp = YAID{255, 255, 255, 255, 255, 255, 255, 255}.Timestamp()

const (
	TimeBytes = 5 // TimeBytes specifies how many bytes are used for the timestamp
	DiffBytes = 2 // DiffBytes specifies how many bytes are used for the differentiator
	MetaBytes = 1 // MetaBytes specifies how many bytes are used for metadata

	// Defider of the timestamp. 1 for miliseconds, 10 for centiseconds, 1000 for seconds, etc.
	Defider = 10
)

type YAID [TimeBytes + DiffBytes + MetaBytes]byte

// Return the
func (y YAID) Int() uint64 {
	return binary.BigEndian.Uint64(y[:])
}

// Return id as Base32 crockford encoded string
func (y YAID) String() string {
	return crock32.Encode(y[:])
}

func (y *YAID) Differentiator() []byte {
	return y[TimeBytes : TimeBytes+DiffBytes]
}

// Set the differentiator
func (y *YAID) SetDifferentiator(b []byte) error {

	if len(b) != DiffBytes {
		return fmt.Errorf("random part must be exactly %d bytes", DiffBytes)
	}

	for i := 0; i < DiffBytes; i++ {
		(*y)[i+TimeBytes] = b[i]
	}

	return nil
}

func (y *YAID) Meta() []byte {
	return y[TimeBytes+DiffBytes:]
}

func (y *YAID) SetMeta(meta []byte) error {
	if len(meta) > MetaBytes {
		return fmt.Errorf("meta key must not be longer than %d bytes", MetaBytes)
	}

	for i := 0; i < MetaBytes; i++ {
		(*y)[i+TimeBytes+DiffBytes] = meta[i]
	}

	return nil
}

func (y YAID) Time() time.Time {
	ms := y.Timestamp() * Defider
	return time.UnixMilli(int64(ms))
}

func (y *YAID) SetTime(t time.Time) error {
	ms := t.UnixMilli() / Defider
	return y.SetTimestamp(uint64(ms))
}

func (y YAID) MarshalText() ([]byte, error) {
	return []byte(y.String()), nil
}

func (y *YAID) UnmarshalText(text []byte) (err error) {
	*y, err = Parse(string(text))
	return err
}

// Returns the time as hundredth of a second since January 1, 1970 12:00:00 AM UTC
func (y YAID) Timestamp() uint64 {
	return uint64(y[4]) |
		uint64(y[3])<<8 |
		uint64(y[2])<<16 |
		uint64(y[1])<<24 |
		uint64(y[0])<<32
}

// Set the time as hundredth of a second since January 1, 1970 12:00:00 AM UTC
func (y *YAID) SetTimestamp(t uint64) error {
	if t > maxTimestamp {
		return fmt.Errorf("epoch must not be greater than %d", maxTimestamp)
	}

	(*y)[0] = byte(t >> 32)
	(*y)[1] = byte(t >> 24)
	(*y)[2] = byte(t >> 16)
	(*y)[3] = byte(t >> 8)
	(*y)[4] = byte(t)

	return nil
}

type Generator struct {
	Meta   []byte
	Random io.Reader
}

func (g Generator) New() (y YAID, err error) {
	err = y.SetTime(time.Now())
	if err != nil {
		return y, err
	}

	err = y.SetMeta(g.Meta)
	if err != nil {
		return y, err
	}

	b := make([]byte, DiffBytes)
	_, err = g.Random.Read(b)
	if err != nil {
		return y, err
	}
	y.SetDifferentiator(b)

	return y, err
}

// Create a factory that returns ids.
func Factory(meta []byte) func() (y YAID, err error) {
	g := Generator{meta, rand.Reader}
	return func() (y YAID, err error) {
		return g.New()
	}
}

func New(meta []byte) (y YAID, err error) {
	return Factory(meta)()
}

func Parse(yaid string) (y YAID, err error) {
	b, err := crock32.Decode(yaid)
	if err != nil {
		return y, err
	}
	if len(b) != TimeBytes+DiffBytes+MetaBytes {
		return y, fmt.Errorf("yaid must be exactly %d bytes", TimeBytes+DiffBytes+MetaBytes)
	}

	copy(y[:], b)

	return y, nil
}
